enhancements/ingress/global-admission-hsts.md

```diff
@@ -0,0 +1,726 @@
+        case DefaultPreloadPolicy, NoOpinionPreloadPolicy:
+        // anything is allowed, do nothing
+        case RequirePreloadPolicy:
+            if !c.preload {
+                return fmt.Errorf("preload must be true")
+            }
+        case RequireNoPreloadPolicy:
+            if c.preload {
+                return fmt.Errorf("preload must be unspecified")
+            }
+    }   
+
+    switch requirement.IncludeSubdomainsPolicy {
+        case DefaultIncludeSubdomains, NoOpinionIncludeSubdomains:
+        // anything is allowed, do nothing
+        case RequireIncludeSubdomains:
+            if !c.includeSubdomains {
+                return fmt.Errorf("includeSubdomains must be true")
+            }
+        case RequireNotIncludeSubdomains:
+            if c.includeSubdomains {
+                return fmt.Errorf("includeSubdomains must be unspecified")
+            }
+    }
+
+    return nil
+}
+// Check if the route matches the required domain/namespace in the HSTS Policy
+func requiredHSTSMatchesRoute(requirement RequiredHSTSPolicy, route *routeapi.Route, namespace *corev1.Namespace) (bool, error) {
+    matchesNamespace, err := matchesNamespaceSelector(requirement.NamespaceSelector, namespace)
+    if err != nil {
+        return false, err
+    }
+    if !matchesNamespace {
+        return false, nil
+    }
+    routeDomains := []string{route.Spec.Host}
+    for _, ingress := range route.Status.Ingress {
+        routeDomains = append(routeDomains, ingress.Host)
+    }
+    if matchesDomain(requirement.DomainPatterns, routeDomains) {
+        return true, nil
+    }
+
+    return false, nil
+}
+````
+
+Formatting of the HSTS Annotation field on a `route`:
+```yaml
+max-age=<numberofseconds>[;includeSubDomains][;preload]`
+```
+
+The `RequiredHSTSPolicies` variable can be configured by the administrator in the `Ingress`, as shown in this example, where
+there is one HSTS Policy is for domain abc.com, expires after one year, includes subdomains, and requests preload:
+```yaml
+apiVersion: config.openshift.io/v1
+kind: Ingress
+metadata:
+  name: cluster
+spec:
+  domain: apps.abc.com
+  requiredHSTSPolicies:
+    - domainPatterns:
+        - abc.com
+      maxAge:
+          smallestMaxAge: 1
+          largestMaxAge: 31536000
+      preloadPolicy: "RequirePreload"
+      includeSubdomainsPolicy: "RequireIncludeSubdomains"
+```
+
+#### Route Administration
+
+To handle upgraded clusters with non-compliant HSTS routes, the best solution is to update
+the manifests at the source and apply the updates.  However, in upgrade clusters, administrators
+can update the routes directly via the API.
+
+To apply HSTS to all routes in the cluster, use a command like this. (This command requires
+an updated API that supports the `--all-namespaces` flag on the `oc annotate` verb):
+```shell
+$ oc annotate route --all --all-namespaces --overwrite=true "haproxy.router.openshift.io/hsts_header"="max-age=31536000"
```

Comment:
> This command only makes sense if all routes are TLS routes. Is that typical?

I'm fairly sure the command continues through all even after encountering errors, so I think it would still work.  However, is there an easy way to filter on whether the route is http or https using a field selector?

Reply:
Good point.  It will fail on non-TLS routes.  There are other ways listed to perform annotations.

Reply:
I changed this so it doesn't fail on non-TLS routes, per https://github.com/openshift/enhancements/pull/749/files#r652922363